### ISSUE TITLE
Make batch insertion of multiple sets more intelligent through topological ordering

### ISSUE THREAD

**Issue by ymotton** [MikaelEliasson/EntityFramework.Utilities#40]
When writing data import for large sets of data I often find myself writing something along the lines of:

    var sourceOrders = sourceContext.Orders.ToList();
    var mappedOrders = MapToTarget(sourceOrders);
    targetContext.Orders.AddRange(mappedOrders);
    targetContext.SaveChanges();

Where MapToTarget maps the top level entity (Order) and all the nested entities underneath (for instance, Customer, OrderLine, Product, ...). In the end you AddRange and SaveChanges, and expect EF to graciously handle the import operation of a couple of thousand entities.

I don't need to tell you how wrong the above assumption is.

One of the ways BatchInsert/BulkInsert can help here, is that if you know the dependency structure of the Entities, you can flatten the graph into 4 sets: Customers, Orders, Products and OrderLines, and insert them in the reverse order of dependency. Root to leaf node.

    var sourceOrders = sourceContext.Orders.ToList();
    var mappedOrders = MapToTarget(sourceOrders);
    EFBatchOperation.For(targetContext, targetContext.Customers).InsertAll(mappedOrders.Select(o => o.Customer));
    EFBatchOperation.For(targetContext, targetContext.Orders).InsertAll(mappedOrders);
    EFBatchOperation.For(targetContext, targetContext.Products).InsertAll(mappedOrders.SelectMany(o => o.OrderLines.Select(ol => ol.Product)));
    EFBatchOperation.For(targetContext, targetContext.OrderLines).InsertAll(mappedOrders.SelectMany(o => o.OrderLines));

I have found that in complex object models where the dependency structure might not be super clear, it can help to have a library assist you in topologically sorting the added entity sets by their dependency, and then bulk inserting the ordered sets.

The idea is that given that you have an acyclic dependency graph between your entities, you can leverage the EF changetracker to do the heavy lifting in terms of figuring out what has changed, and what the dependencies are between the changes, in order to then save the changes via bulkinsert:

    // Just add them to the "Orders" DbSet, and have EF handle figuring out what has changed
    targetContext.Orders.AddRange(mappedOrders);
    // ... then, using the ObjectStateManager, and the metadata we have about the dependency graph, sort, and InsertAll
    targetContext.SaveAdditions();

SaveAdditions would first verify that the ObjectStateManager only contains added entities (not modified or deleted ones).
Then you would topologically sort based on their dependency.
Then you could bulk insert in decending order of dependency: Customer > Order > Product > OrderLine.

I have an implementation of this based on a different BulkInsert package: https://github.com/ymotton/EntityFramework.BulkExtensions/blob/master/EntityFramework.BulkExtensions/DbContextBulkExtensions.cs

**Comment by MikaelEliasson** [comment]
Very interesting. How do you find the performance of this solution? The changetracker can be quite slow but i guess it should be barely noticable with a single DetectChanges cycle.

I guess you have application generated id's for all entities here?

**Comment by ymotton** [comment]
Well, the performance was marginally worse than handrolling the Batch Insert statements in the correct order yourself. And obviously, you can't even compare to vanilla SaveChanges().

And yeah, application generated id's are a must. Does SqlBlkCopy support store generated ids?

**Comment by MikaelEliasson** [comment]
Yes, But you don't get the id back so you can't populate the fks with it.

I've been thinking about a way to allow that with temp tables and a temporary id. Shouldn't affect this too much anyway

**Comment by MikaelEliasson** [comment]
I got Id return working with SqlBulkCopy today. That means it will be possible to make bulk insert of related entities work even with store generated id's. It's about 2x slower but compared to the alternatives that's still uncomparable.

It will be interesting to try this out next.
